File: pz-mod/42/media/lua/shared/PZAgent/Refs.lua
--[[
PZAgent.Refs -- stable, session-scoped references to game objects.

Invariant: the strings this module builds and parses are byte-identical to the
ones produced by pz_agent_core.protocol.refs. The two sides never exchange a
Lua table or a Java handle, only these strings, so a disagreement about where
the delimiters fall does not fail loudly -- it silently resolves to a different
object, which is exactly the class of bug the whole reference scheme exists to
prevent.

The subtle part is the item reference. Its container tail itself contains
colons ("world:10726:9455:0:3:1"), so an item reference must be split from the
ends -- last colon for the generation, next-to-last for the runtime id, and
everything between the session and those two is the tail, verbatim. A naive
left-to-right split produces a reference that looks plausible and points at the
wrong container.

Every parser returns `table` or `nil, message`. Nothing raises: a malformed
reference is a routine event on a boundary that accepts input from outside the
game.
]]

PZAgent = PZAgent or {}

local Refs = {}
PZAgent.Refs = Refs

Refs.SEPARATOR = ":"

Refs.KIND = {
  ITEM = "item",
  CONTAINER = "container",
  SQUARE = "square",
  ZOMBIE = "zombie",
  WOUND = "wound",
  OBJECT = "object",
}

local KIND_SET = {}
for _, kind in pairs(Refs.KIND) do
  KIND_SET[kind] = true
end

--- No legitimate reference comes close to this. The bound keeps the scanning
--- loops below finite on input that arrived from the exchange directory.
local MAX_REF_BYTES = 1024

--- Container and item segments use a restricted alphabet so a reference can
--- never smuggle a path separator or a delimiter. Mirrors _SAFE_SEGMENT.
local SEGMENT_PATTERN = "^[A-Za-z0-9_%.%-]+$"

--- Mirrors _SESSION_SEGMENT: UUIDs, and nothing that could contain a colon.
local SESSION_PATTERN = "^[A-Za-z0-9%-]+$"
local MAX_SESSION_BYTES = 64

local function isSegment(value)
  return type(value) == "string" and value:match(SEGMENT_PATTERN) ~= nil
end

local function isSession(value)
  return type(value) == "string"
    and #value >= 1
    and #value <= MAX_SESSION_BYTES
    and value:match(SESSION_PATTERN) ~= nil
end

Refs.isSessionSegment = isSession
Refs.isSafeSegment = isSegment

--- Split at the first separator. Mirrors Python's str.partition: when there is
--- no separator the whole string is the head and the tail is empty.
local function partition(text)
  local index = text:find(Refs.SEPARATOR, 1, true)
  if index == nil then
    return text, false, ""
  end
  return text:sub(1, index - 1), true, text:sub(index + 1)
end

--- Split at the last separator. Mirrors Python's str.rpartition: when there is
--- no separator the head is empty and the whole string is the tail.
local function rpartition(text)
  local last = nil
  local cursor = 1
  while true do
    local index = text:find(Refs.SEPARATOR, cursor, true)
    if index == nil then
      break
    end
    last = index
    cursor = index + 1
  end
  if last == nil then
    return "", false, text
  end
  return text:sub(1, last - 1), true, text:sub(last + 1)
end

--- Accept only what SquareRef/ZombieRef would have written: an optionally
--- signed run of ASCII digits. Deliberately stricter than Python's int(),
--- which also accepts surrounding whitespace and underscore separators --
--- neither can be produced by the writer, so both mean tampering here.
local function toInteger(text)
  if type(text) ~= "string" or text:match("^[-+]?%d+$") == nil then
    return nil
  end
  return tonumber(text)
end

local function isCount(text)
  return type(text) == "string" and text:match("^%d+$") ~= nil
end

local function checkRaw(raw)
  if type(raw) ~= "string" then
    return "reference must be a string, got " .. type(raw)
  end
  if #raw == 0 then
    return "reference is empty"
  end
  if #raw > MAX_REF_BYTES then
    return string.format("reference of %d bytes exceeds the %d byte limit", #raw, MAX_REF_BYTES)
  end
  return nil
end

-- ---------------------------------------------------------------------------
-- container references
-- ---------------------------------------------------------------------------

--- container:<session>:<tail>
function Refs.buildContainer(sessionId, tail)
  if not isSession(sessionId) then
    return nil, "invalid session segment"
  end
  if type(tail) ~= "string" or #tail == 0 then
    return nil, "container tail is empty"
  end
  for part in (tail .. Refs.SEPARATOR):gmatch("([^:]*)" .. Refs.SEPARATOR) do
    if not isSegment(part) then
      return nil, string.format("invalid container tail segment: %q", part)
    end
  end
  return Refs.KIND.CONTAINER .. Refs.SEPARATOR .. sessionId .. Refs.SEPARATOR .. tail
end

function Refs.playerMainContainer(sessionId)
  return Refs.buildContainer(sessionId, "player-main")
end

function Refs.wornContainer(sessionId, slot, itemRuntimeId)
  if not isSegment(slot) then
    return nil, "invalid slot segment"
  end
  if not isSegment(itemRuntimeId) then
    return nil, "invalid runtime id segment"
  end
  return Refs.buildContainer(sessionId, "worn" .. Refs.SEPARATOR .. slot .. Refs.SEPARATOR .. itemRuntimeId)
end

function Refs.carriedContainer(sessionId, itemRuntimeId)
  if not isSegment(itemRuntimeId) then
    return nil, "invalid runtime id segment"
  end
  return Refs.buildContainer(sessionId, "carried" .. Refs.SEPARATOR .. itemRuntimeId)
end

--- container:<session>:world:<x>:<y>:<z>:<object-index>:<container-index>
function Refs.worldContainer(sessionId, x, y, z, objectIndex, containerIndex)
  local parts = { "world" }
  local coordinates = { x, y, z, objectIndex, containerIndex }
  for index = 1, 5 do
    local value = coordinates[index]
    if type(value) ~= "number" or value ~= math.floor(value) then
      return nil, "world container coordinates must be integers"
    end
    parts[index + 1] = string.format("%.0f", value)
  end
  return Refs.buildContainer(sessionId, table.concat(parts, Refs.SEPARATOR))
end

function Refs.parseContainer(raw)
  local rawError = checkRaw(raw)
  if rawError then
    return nil, rawError
  end
  local kind, _, rest = partition(raw)
  if kind ~= Refs.KIND.CONTAINER then
    return nil, "not a container ref"
  end
  local session, _, tail = partition(rest)
  if #tail == 0 then
    return nil, "container ref missing tail"
  end
  if not isSession(session) then
    return nil, "invalid session segment"
  end
  for part in (tail .. Refs.SEPARATOR):gmatch("([^:]*)" .. Refs.SEPARATOR) do
    if not isSegment(part) then
      return nil, string.format("invalid container tail segment: %q", part)
    end
  end
  return {
    kind = Refs.KIND.CONTAINER,
    session_id = session,
    tail = tail,
  }
end

function Refs.isPlayerMainTail(tail)
  return tail == "player-main"
end

--- True for the main inventory and anything worn or carried -- the containers
--- that move with the player and so need no travel to reach.
function Refs.isOnPersonTail(tail)
  if type(tail) ~= "string" then
    return false
  end
  return tail == "player-main" or tail:sub(1, 5) == "worn:" or tail:sub(1, 8) == "carried:"
end

function Refs.isWorldTail(tail)
  return type(tail) == "string" and tail:sub(1, 6) == "world:"
end

-- ---------------------------------------------------------------------------
-- item references
-- ---------------------------------------------------------------------------

--- item:<session>:<container-tail>:<runtime-id>:<generation>
function Refs.buildItem(sessionId, containerTail, runtimeId, generation)
  if not isSession(sessionId) then
    return nil, "invalid session segment"
  end
  if type(containerTail) ~= "string" or #containerTail == 0 then
    return nil, "container tail is empty"
  end
  if not isSegment(runtimeId) then
    return nil, "invalid runtime id segment"
  end
  if type(generation) ~= "number" or generation ~= math.floor(generation) or generation < 0 then
    return nil, "generation must be a non-negative integer"
  end
  return table.concat({
    Refs.KIND.ITEM,
    sessionId,
    containerTail,
    runtimeId,
    string.format("%.0f", generation),
  }, Refs.SEPARATOR)
end

--- Parse an item reference from the ends, never left to right.
---
--- The container tail is returned verbatim and is not re-validated segment by
--- segment: ItemRef.parse on the Python side does not validate it either, and
--- a mod that were stricter would reject a reference the sidecar considers
--- well-formed. Callers that need the tail checked run it through
--- parseContainer or containerRefOf.
function Refs.parseItem(raw)
  local rawError = checkRaw(raw)
  if rawError then
    return nil, rawError
  end
  local kind, _, rest = partition(raw)
  if kind ~= Refs.KIND.ITEM then
    return nil, "not an item ref"
  end
  local session, _, remainder = partition(rest)
  local head, hasGeneration, generation = rpartition(remainder)
  if not hasGeneration then
    return nil, "item ref missing generation"
  end
  local containerTail, hasRuntimeId, runtimeId = rpartition(head)
  if not hasRuntimeId then
    return nil, "item ref missing runtime id"
  end
  if not isCount(generation) then
    return nil, "item ref generation must be numeric"
  end
  if not isSession(session) then
    return nil, "invalid session segment"
  end
  if not isSegment(runtimeId) then
    return nil, "invalid runtime id segment"
  end
  return {
    kind = Refs.KIND.ITEM,
    session_id = session,
    container_tail = containerTail,
    runtime_id = runtimeId,
    generation = tonumber(generation),
  }
end

--- Reconstruct the owning container reference of a parsed item.
function Refs.containerRefOf(item)
  if type(item) ~= "table" then
    return nil, "expected a parsed item ref"
  end
  return Refs.buildContainer(item.session_id, item.container_tail)
end

-- ---------------------------------------------------------------------------
-- square and zombie references
-- ---------------------------------------------------------------------------

--- square:<session>:<x>:<y>:<z>
function Refs.buildSquare(sessionId, x, y, z)
  if not isSession(sessionId) then
    return nil, "invalid session segment"
  end
  local parts = { Refs.KIND.SQUARE, sessionId }
  local coordinates = { x, y, z }
  for index = 1, 3 do
    local value = coordinates[index]
    if type(value) ~= "number" or value ~= math.floor(value) then
      return nil, "square coordinates must be integers"
    end
    parts[index + 2] = string.format("%.0f", value)
  end
  return table.concat(parts, Refs.SEPARATOR)
end

function Refs.parseSquare(raw)
  local rawError = checkRaw(raw)
  if rawError then
    return nil, rawError
  end
  local parts = {}
  local count = 0
  for part in (raw .. Refs.SEPARATOR):gmatch("([^:]*)" .. Refs.SEPARATOR) do
    count = count + 1
    if count > 5 then
      return nil, "not a square ref"
    end
    parts[count] = part
  end
  if count ~= 5 or parts[1] ~= Refs.KIND.SQUARE then
    return nil, "not a square ref"
  end
  if not isSession(parts[2]) then
    return nil, "invalid session segment"
  end
  local x, y, z = toInteger(parts[3]), toInteger(parts[4]), toInteger(parts[5])
  if x == nil or y == nil or z == nil then
    return nil, "square ref has non-integer coordinate"
  end
  return { kind = Refs.KIND.SQUARE, session_id = parts[2], x = x, y = y, z = z }
end

--- Grid distance ignoring the floor; movement checks compare floors separately.
function Refs.chebyshevDistance(first, second)
  return math.max(math.abs(first.x - second.x), math.abs(first.y - second.y))
end

--- zombie:<session>:<runtime-id>:<generation>
function Refs.buildZombie(sessionId, runtimeId, generation)
  if not isSession(sessionId) then
    return nil, "invalid session segment"
  end
  if not isSegment(runtimeId) then
    return nil, "invalid runtime id segment"
  end
  if type(generation) ~= "number" or generation ~= math.floor(generation) or generation < 0 then
    return nil, "generation must be a non-negative integer"
  end
  return table.concat({
    Refs.KIND.ZOMBIE,
    sessionId,
    runtimeId,
    string.format("%.0f", generation),
  }, Refs.SEPARATOR)
end

function Refs.parseZombie(raw)
  local rawError = checkRaw(raw)
  if rawError then
    return nil, rawError
  end
  local parts = {}
  local count = 0
  for part in (raw .. Refs.SEPARATOR):gmatch("([^:]*)" .. Refs.SEPARATOR) do
    count = count + 1
    if count > 4 then
      return nil, "not a zombie ref"
    end
    parts[count] = part
  end
  if count ~= 4 or parts[1] ~= Refs.KIND.ZOMBIE then
    return nil, "not a zombie ref"
  end
  if not isCount(parts[4]) then
    return nil, "zombie ref generation must be numeric"
  end
  if not isSession(parts[2]) then
    return nil, "invalid session segment"
  end
  if not isSegment(parts[3]) then
    return nil, "invalid runtime id segment"
  end
  return {
    kind = Refs.KIND.ZOMBIE,
    session_id = parts[2],
    runtime_id = parts[3],
    generation = tonumber(parts[4]),
  }
end

-- ---------------------------------------------------------------------------
-- generic inspection
-- ---------------------------------------------------------------------------

--- Classify a reference without fully parsing it.
function Refs.kindOf(raw)
  if type(raw) ~= "string" then
    return nil, "reference must be a string"
  end
  local kind = partition(raw)
  if not KIND_SET[kind] then
    return nil, "unknown ref kind"
  end
  return kind
end

--- True when `raw` was minted by `sessionId`.
---
--- A reference from an earlier session is not merely stale: its runtime ids may
--- now denote entirely different objects, so a false here is INVALID_REF and
--- never a reason to retry.
function Refs.belongsToSession(raw, sessionId)
  if type(raw) ~= "string" or type(sessionId) ~= "string" then
    return false
  end
  local kind, _, rest = partition(raw)
  if not KIND_SET[kind] then
    return false
  end
  local candidate = partition(rest)
  return candidate == sessionId
end

--- Parse any reference kind this module understands.
function Refs.parse(raw)
  local kind, kindError = Refs.kindOf(raw)
  if kind == nil then
    return nil, kindError
  end
  if kind == Refs.KIND.ITEM then
    return Refs.parseItem(raw)
  elseif kind == Refs.KIND.CONTAINER then
    return Refs.parseContainer(raw)
  elseif kind == Refs.KIND.SQUARE then
    return Refs.parseSquare(raw)
  elseif kind == Refs.KIND.ZOMBIE then
    return Refs.parseZombie(raw)
  end
  return nil, string.format("no parser for ref kind %q", kind)
end

return Refs

File: packages/pz_agent_core/src/pz_agent_core/protocol/refs.py
"""Stable, session-scoped references to game objects.

The LLM never sees a Lua table or a Java handle. It sees an opaque string that
the mod can re-resolve and re-validate immediately before acting on it
(docs/blueprint/03_GAME_STATE_AND_IPC.md §3.7).

A reference carries a *generation*. Anything that invalidates object identity —
a save/load transition, a new session — bumps the generation, so a reference
minted before the transition fails validation instead of silently pointing at a
different object.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import StrEnum
from typing import Final

#: Session ids are UUIDs; container/item segments are restricted to a safe
#: alphabet so a reference can never smuggle a path separator or a delimiter.
_SAFE_SEGMENT: Final = re.compile(r"^[A-Za-z0-9_.\-]+$")
_SESSION_SEGMENT: Final = re.compile(r"^[A-Za-z0-9\-]{1,64}$")

REF_SEPARATOR: Final = ":"


class RefKind(StrEnum):
    """Leading token of a reference string."""

    ITEM = "item"
    CONTAINER = "container"
    SQUARE = "square"
    ZOMBIE = "zombie"
    WOUND = "wound"
    OBJECT = "object"


class RefError(ValueError):
    """A reference is malformed, or belongs to a different session."""


def _check_segment(value: str, *, field: str) -> str:
    if not value or not _SAFE_SEGMENT.match(value):
        raise RefError(f"invalid {field} segment: {value!r}")
    return value


def _check_session(value: str) -> str:
    if not _SESSION_SEGMENT.match(value):
        raise RefError(f"invalid session segment: {value!r}")
    return value


@dataclass(frozen=True, slots=True)
class ItemRef:
    """``item:<session>:<container-ref-tail>:<runtime-id>:<generation>``.

    The container tail is stored joined because a container reference itself
    contains colons; splitting is done from both ends, never by a naive
    ``split(":")``.
    """

    session_id: str
    container_tail: str
    runtime_id: str
    generation: int

    def __str__(self) -> str:
        return REF_SEPARATOR.join(
            (
                RefKind.ITEM.value,
                self.session_id,
                self.container_tail,
                self.runtime_id,
                str(self.generation),
            )
        )

    @classmethod
    def parse(cls, raw: str) -> ItemRef:
        kind, _, rest = raw.partition(REF_SEPARATOR)
        if kind != RefKind.ITEM.value:
            raise RefError(f"not an item ref: {raw!r}")
        session, _, rest = rest.partition(REF_SEPARATOR)
        head, sep, generation = rest.rpartition(REF_SEPARATOR)
        if not sep:
            raise RefError(f"item ref missing generation: {raw!r}")
        container_tail, sep, runtime_id = head.rpartition(REF_SEPARATOR)
        if not sep:
            raise RefError(f"item ref missing runtime id: {raw!r}")
        if not generation.isdigit():
            raise RefError(f"item ref generation must be numeric: {raw!r}")
        return cls(
            session_id=_check_session(session),
            container_tail=container_tail,
            runtime_id=_check_segment(runtime_id, field="runtime id"),
            generation=int(generation),
        )

    @property
    def container_ref(self) -> str:
        """The owning container reference, reconstructed."""
        return REF_SEPARATOR.join((RefKind.CONTAINER.value, self.session_id, self.container_tail))


@dataclass(frozen=True, slots=True)
class ContainerRef:
    """``container:<session>:<tail>`` where the tail encodes the location.

    Recognised tails:

    * ``player-main``
    * ``worn:<slot>:<item-runtime-id>``
    * ``carried:<item-runtime-id>``
    * ``world:<x>:<y>:<z>:<object-index>:<container-index>``
    """

    session_id: str
    tail: str

    def __str__(self) -> str:
        return REF_SEPARATOR.join((RefKind.CONTAINER.value, self.session_id, self.tail))

    @classmethod
    def parse(cls, raw: str) -> ContainerRef:
        kind, _, rest = raw.partition(REF_SEPARATOR)
        if kind != RefKind.CONTAINER.value:
            raise RefError(f"not a container ref: {raw!r}")
        session, _, tail = rest.partition(REF_SEPARATOR)
        if not tail:
            raise RefError(f"container ref missing tail: {raw!r}")
        for part in tail.split(REF_SEPARATOR):
            _check_segment(part, field="container tail")
        return cls(session_id=_check_session(session), tail=tail)

    @classmethod
    def player_main(cls, session_id: str) -> ContainerRef:
        return cls(session_id=_check_session(session_id), tail="player-main")

    @classmethod
    def worn(cls, session_id: str, slot: str, item_runtime_id: str) -> ContainerRef:
        return cls(
            session_id=_check_session(session_id),
            tail=REF_SEPARATOR.join(
                (
                    "worn",
                    _check_segment(slot, field="slot"),
                    _check_segment(item_runtime_id, field="runtime id"),
                )
            ),
        )

    @classmethod
    def world(
        cls,
        session_id: str,
        *,
        x: int,
        y: int,
        z: int,
        object_index: int,
        container_index: int,
    ) -> ContainerRef:
        return cls(
            session_id=_check_session(session_id),
            tail=REF_SEPARATOR.join(
                ("world", str(x), str(y), str(z), str(object_index), str(container_index))
            ),
        )

    @property
    def is_player_main(self) -> bool:
        return self.tail == "player-main"

    @property
    def is_on_person(self) -> bool:
        """True for the main inventory and anything worn or carried."""
        return self.is_player_main or self.tail.startswith(("worn:", "carried:"))

    @property
    def is_world(self) -> bool:
        return self.tail.startswith("world:")


@dataclass(frozen=True, slots=True)
class SquareRef:
    """``square:<session>:<x>:<y>:<z>`` — a single grid cell."""

    session_id: str
    x: int
    y: int
    z: int

    def __str__(self) -> str:
        return REF_SEPARATOR.join(
            (RefKind.SQUARE.value, self.session_id, str(self.x), str(self.y), str(self.z))
        )

    @classmethod
    def parse(cls, raw: str) -> SquareRef:
        parts = raw.split(REF_SEPARATOR)
        if len(parts) != 5 or parts[0] != RefKind.SQUARE.value:
            raise RefError(f"not a square ref: {raw!r}")
        _, session, sx, sy, sz = parts
        try:
            return cls(session_id=_check_session(session), x=int(sx), y=int(sy), z=int(sz))
        except ValueError as exc:
            raise RefError(f"square ref has non-integer coordinate: {raw!r}") from exc

    def chebyshev_distance(self, other: SquareRef) -> int:
        """Grid distance ignoring floor; movement checks compare floors separately."""
        return max(abs(self.x - other.x), abs(self.y - other.y))


@dataclass(frozen=True, slots=True)
class ZombieRef:
    """``zombie:<session>:<runtime-id>:<generation>``."""

    session_id: str
    runtime_id: str
    generation: int

    def __str__(self) -> str:
        return REF_SEPARATOR.join(
            (
                RefKind.ZOMBIE.value,
                self.session_id,
                self.runtime_id,
                str(self.generation),
            )
        )

    @classmethod
    def parse(cls, raw: str) -> ZombieRef:
        parts = raw.split(REF_SEPARATOR)
        if len(parts) != 4 or parts[0] != RefKind.ZOMBIE.value:
            raise RefError(f"not a zombie ref: {raw!r}")
        _, session, runtime_id, generation = parts
        if not generation.isdigit():
            raise RefError(f"zombie ref generation must be numeric: {raw!r}")
        return cls(
            session_id=_check_session(session),
            runtime_id=_check_segment(runtime_id, field="runtime id"),
            generation=int(generation),
        )


def ref_kind(raw: str) -> RefKind:
    """Classify a reference string without fully parsing it.

    Raises:
        RefError: if the leading token is not a known kind.
    """
    head, _, _ = raw.partition(REF_SEPARATOR)
    try:
        return RefKind(head)
    except ValueError as exc:
        raise RefError(f"unknown ref kind: {raw!r}") from exc


def belongs_to_session(raw: str, session_id: str) -> bool:
    """True when *raw* was minted by *session_id*.

    A reference from an earlier session is not merely stale — its runtime ids
    may now denote entirely different objects, so callers must treat a False
    here as ``INVALID_REF`` rather than retrying.
    """
    head, _, rest = raw.partition(REF_SEPARATOR)
    if head not in {k.value for k in RefKind}:
        return False
    candidate, _, _ = rest.partition(REF_SEPARATOR)
    return candidate == session_id
